Remora.Discord.Commands/Conditions/RequireUserGuildPermissionCondition.cs

```diff
@@ -0,0 +1,141 @@
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Remora.Commands.Conditions;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Abstractions.Rest;
+using Remora.Discord.API.Objects;
+using Remora.Discord.Commands.Contexts;
+using Remora.Discord.Commands.Results;
+using Remora.Results;
+
+namespace Remora.Discord.Commands.Conditions
+{
+    /// <summary>
+    /// Checks required Guild permissions before allowing execution.
+    ///
+    /// <remarks>Fails if the command is executed outside of a Guild. It should be used together with <see cref=""./></remarks>
+    /// </summary>
+    public class RequireUserGuildPermissionCondition : ICondition<RequireUserGuildPermissionAttribute>
+    {
+        private readonly ICommandContext _context;
+        private readonly IDiscordRestChannelAPI _channelApi;
+        private readonly IDiscordRestGuildAPI _guildApi;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RequireUserGuildPermissionCondition"/> class.
+        /// </summary>
+        /// <param name="context">The command context.</param>
+        /// <param name="guildApi">The guild API.</param>
+        /// <param name="channelApi">The channel API.</param>
+        public RequireUserGuildPermissionCondition(ICommandContext context, IDiscordRestGuildAPI guildApi, IDiscordRestChannelAPI channelApi)
+        {
+            _context = context;
+            _guildApi = guildApi;
+            _channelApi = channelApi;
+        }
+
+        /// <inheritdoc />
+        public async ValueTask<Result> CheckAsync(RequireUserGuildPermissionAttribute attribute, CancellationToken ct)
+        {
+            var getChannel = await _channelApi.GetChannelAsync(_context.ChannelID, ct);
+            if (!getChannel.IsSuccess)
+            {
+                return Result.FromError(getChannel);
+            }
+            var channel = getChannel.Entity;
+            if (!channel.GuildID.HasValue)
+            {
+                return new ConditionNotSatisfiedError("Command requires a guild permission but was executed outside of a guild.");
+            }
+
+            var guildId = channel.GuildID.Value;
+
+            var getGuildMember = await _guildApi.GetGuildMemberAsync(guildId, _context.User.ID, ct);
+            if (!getGuildMember.IsSuccess)
+            {
+                return Result.FromError(getGuildMember);
+            }
+
+            var getGuildRoles = await _guildApi.GetGuildRolesAsync(guildId, ct);
+            if (!getGuildRoles.IsSuccess)
+            {
+                return Result.FromError(getGuildRoles);
+            }
+
+            var guildRoles = getGuildRoles.Entity;
+            var everyoneRole = guildRoles.FirstOrDefault(r => r.Name.Equals("@everyone"));
```

Comment:
You should probably check for `guild.ID == role.ID` instead - it's a bit more robust in this case.